### DISCUSSION TITLE
Support for WEG SIW200G inverter?

### DISCUSSION THREAD

**Discussion by LucasTor** [nathanmarlor/foxess_modbus#439] [Q&A] [answered]
I know SIW200G inverters are made by FoxESS, and the FoxESS app can be used to connect/control them, but I did not find the exact FoxESS model that would represent this inverter, do you guys know if this integration can communicate to it via modbus?
[This is the inverter in question](https://www.weg.net/catalog/weg/BR/pt/Automa%C3%A7%C3%A3o-e-Controle-Industrial/Drives/M%C3%B3dulos-e-Inversores-Fotovoltaicos/Inversor-Monof%C3%A1sico-SIW200G/Inversor-Monof%C3%A1sico-SIW200G/p/MKT_WDC_BRAZIL_INVERTER1PHASE_SIW200)
Thanks!!

**Comment by canton7** [comment]
I've no idea I'm afraid. I can't even find an installation manual which would show what connections are available on the bottom (although I'm on mobile and didn't look too hard).

If you've actually got one, the thing to do would be to see if there's an ethernet socket / RS485 connector you can connect to, and start probing it. 

If not, we'd need to know what foxess model it was similar to. Only some of the foxess inverters talk modbus - the T series just spits out information over serial for example.

**Comment by LucasTor** [comment]
I still don't have one, the company I bought it from gave me a 60 day deadline to install it, but I checked the manual for it, and it has a RS485 port on the bottom, which I plan on using with an EW11A to send data to Home Assistant (I would use USB but it's too far from my server)
I was able to find it's protocol online and it's the same as the SIW400G, which is the FoxESS T10 model exactly, so should I assume I can use this integration as mine has the same protocol as the T10?

**Reply by canton7** [reply, marked as answer]
No, the T series doesn't talk modbus, and isn't supported by this integration. The integration to talk to T series inverters is https://github.com/assembly12/Foxess-T-series-ESPHome-Home-Assistant (although you could write your own to avoid having to go the esphome route: the inverter just spits out information periodically)

**Reply by mattwilson1066** [reply]
I'd love to know how to read the data that the inverter spits out without needing to use the ESPHome option, I have an F4600 inverter (and an H1-6.0-E which is connected modbus and works like a dream). I have wired up to the dongle, so I have a feed from the A/B of the RS485, but I am not sure what i'd need to do as the next step for collecting and displaying the data. I have a https://www.waveshare.com/usb-to-rs485-b.htm  which I planned to connect the A/B connection directly into, but not sure what next...

**Reply by LucasTor** [reply]
@mattwilson1066 I've found the thread in which the esphome project came from, in there there's a guy that setup elfin ew11 with node-red to integrate the inverter data in to home assistant, I'm not sure if node red can read from USB tho, as I never used it
This is the thread where they created the ESPhome project: https://community.home-assistant.io/t/foxess-inverter-modbus/459313/48
This is a "lib" I found that implements USB Serial data with node red: https://flows.nodered.org/node/node-red-contrib-serialport-rs485
Maybe if you merge those 2 things you can read from USB and parse it into Home Assistant, if you are interested in tinkering with node red

**Reply by LucasTor** [reply]
Thanks @canton7, I'll likely go the esphome route as I'm already familiar with it and have a hand full of devices, I appreciate your time

**Reply by canton7** [reply]
@mattwilson1066 How good is your Python? If you're comfortable programming it's a pretty easy job to open the serial device, receive the data, parse it, and publish the various values to a set of sensors (and quite frankly I have no idea why there's a (complex) ESPHome solution but no (simple) solution using a USB device). If you're not comfortable with that, it's probably going to be quite hard I'm afraid.

**Reply by LucasTor** [reply]
@canton7 I might try that route as well, as I work as a web developer, and I feel pretty comfortable with that, and also I already bought an Elfin EW11, so I might as well use that
Your suggestion would be to write a custom integration in Python and then use it inside home assistant? Or running the script "standalone"?
I found this guide that seems to be what we want: https://developers.home-assistant.io/docs/creating_component_index/
If I have any progress on that I'll let you guys know

**Reply by canton7** [reply]
I'd go the route of a proper integration. It'd be pretty simple, and I'm sure other people would thank you for it! Sadly the documentation is rubbish -- I learnt by reading the source of other integrations.

I recently set up a new integration from scratch, and this is the first commit which had everything including HACS working: https://github.com/canton7/solar_battery_forecast/tree/a95976a26403d7cb7f95f0cfb9217b8e98798658 . Feel free to copy! Obviously the stuff in notebooks/ and brains/ isn't relevant, but hopefully the rest provides some inspiration.

The mental model is that `__init__.py` is called when the integration is loaded: this takes a config entry (created during the config flow when the integration is added) creates the controller from it, and stores it in hass. The controller will open the serial device, receive data, parse it, and expose it. Then, later, the entities are created by hass. These register themselves on the controller, which notifies them when new data is available.

Happy to give pointers if you tag me, but as I don't have a T-series it hasn't made sense for me to actually write something myself!

**Reply by LucasTor** [reply]
That's great! That will for sure help a lot, I'm now waiting on the contractor to install the photovoltaic system for me, but I plan on doing some tests beforehand, and getting the home assistant part ready, a few users on that thread posted example strings from the inverter, so I can do the decoding part as well, just leaving me with the communication part once the inverter is installed
Thanks again @canton7

**Reply by LucasTor** [reply]
Hey guys! Some updates on the project:
I was able to decode the inverter data using Python and the struct lib, using this code: 
```python
import struct
from datetime import datetime

hex = "7E7E026323EDAC00E8000017E80000095B0055138707F109530055138707EA09630055138707F814DA006000000E7F001B0000000000000000000000000000002C002C002C00820000555A00000000000000000000000000000000000000000000000000035078001000000000001A00000000000000000000022100020000000000000000000000000000000000000000000000000000000000000000000200020000000003E7FF6417EA000100000000000000000000000000000000000000000000FFFF0002000000000000000000000000000000000000000000000000000000013FE5000000000000000000000000C663E7E7"
result = struct.unpack('>3xi2xhhhhhhhhhhh', bytes.fromhex(hex[0:62]))

(
    timestamp,
    grid_power,
    gen_power,
    load_power,
    grid_voltage_R,
    grid_current_R,
    grid_frequency_R,
    grid_power_R,
    grid_voltage_S,
    grid_current_S,
    grid_frequency_S,
    grid_power_S
) = result

def set_resolution (value, resolution):
    return value / (1 / resolution)

timestamp = datetime.fromtimestamp(timestamp)
grid_voltage_R = set_resolution(grid_voltage_R, 0.1)
grid_current_R = set_resolution(grid_current_R, 0.1)
grid_frequency_R = set_resolution(grid_frequency_R, 0.01)
grid_voltage_S = set_resolution(grid_voltage_S, 0.1)
grid_current_S = set_resolution(grid_current_S, 0.1)
grid_frequency_S = set_resolution(grid_frequency_S, 0.01)

print(timestamp, grid_power, gen_power, load_power, grid_voltage_R, grid_current_R, grid_frequency_R, grid_voltage_S, grid_current_S, grid_frequency_S)
```

I am still missing a few values, as a proof of concept it already works great...
As soon as I have time I will create the repo and add all the home assistant integration boilerplate
Also my Elfin EW11 arrived today :)
<img width="300" alt="Screenshot 2023-10-30 at 13 08 54" src="https://github.com/nathanmarlor/foxess_modbus/assets/25714552/02b00053-7c73-43b2-93f9-ecb5033f40ff">
I don't have the inverter running yet but I guess I can work on trying to connect it to the python script
@mattwilson1066 would you be interested in writing the USB part of the integration? As I only have the EW11, maybe if you know Python you could work on that

**Reply by mattwilson1066** [reply]
Hi LucasTor, frustratingly I have very little coding knowledge, so other than tinkering with code others have written, i'm a long way from being able to write a USB integration. I do work with some guys who are good at Python and I can certainly ask them if they can help.

**Reply by mattwilson1066** [reply]
@LucasTor When i started looking online this was about as far as I got in terms of using the USB to read the serial connection. I already have 1 Waveshare USB to RS485 working for the modbus for my H1, so that should work out the box. I'm not sure if the below is any help or if that is all just standard stuff for people with the skills.

```python
import serial

# Define the serial port and baud rate
serial_port = '/dev/ttyUSB0'  # Change this to your serial port
baud_rate = 9600  # Change this to match your device's baud rate

try:
    # Open the serial connection
    ser = serial.Serial(serial_port, baud_rate)

    # Continuously read data from the serial port
    while True:
        data = ser.readline()  # Read a line of data
        decoded_data = data.decode('utf-8')  # Decode bytes to a string
        print(decoded_data, end='')  # Print the data without a newline

except serial.SerialException as e:
    print(f"An error occurred: {e}")

finally:
    if ser.is_open:
        ser.close()  # Close the serial connection when done

```

**Reply by canton7** [reply]
Fantastic! I'd read through the code for the existing T series integration, see what it's doing with each byte. I did thay in the past at some point, was fairly easy to read.

**Reply by LucasTor** [reply]
Hey guys! Bringing some good news for the project, here is the repo for it: https://github.com/LucasTor/FoxESS-T-series
It wasn't tested with the real inverter yet tho, the only tests I've done were mocking the payload with an Arduino sending to the Elfin EW11, and that worked flawlessly, let's hope everything will go as smoothly when I plug It in the real thing
Thanks again @canton7 for the help with writing the integration, I started with a cookie-cutter template I found online, and used a lot of your code to have some inspiration on how to integrate everything, I am still missing a good code cleanup and refactor, but that will happen once I know everything is working properly
@mattwilson1066 I think if you feel comfortable with it, you can try integrating your USB code in the repo there, likely add an option on the config flow to choose between USB and TCP, and handle the data fetching accordingly, other than that it should all be the same, I'm here to help in case you have any questions :)

**Reply by mattwilson1066** [reply]
Amazing work @LucasTor sadly I have no real coding experience (certainly not at this level) so was very much feeling my way around with the help of ChatGPT and hadn't even got fully into testing the script above. I'm off on holiday tomorrow for just over a week but I am happy to try and a few things when I am back, unless you can find someone who actually knows what they are doing in this side of things (i'm already well out of my depth) but will try when I am back

**Reply by mattwilson1066** [reply]
@LucasTor @canton7 how would I go about integrating the USB code to try and get this up and running? Any pointers greatly appreciated.

**Reply by LucasTor** [reply]
Hey @mattwilson1066, sorry for the delayed response, this end of year has been really busy for me

First thing I would certainly do is get the script you posted above running, and get data from the inverter, what you need to get is a string that starts with 7E7E and ends with E7E7

Here is an example of one:
```7E7E026323EDAC00E8000017E80000095B0055138707F109530055138707EA09630055138707F814DA006000000E7F001B0000000000000000000000000000002C002C002C00820000555A00000000000000000000000000000000000000000000000000035078001000000000001A00000000000000000000022100020000000000000000000000000000000000000000000000000000000000000000000200020000000003E7FF6417EA000100000000000000000000000000000000000000000000FFFF0002000000000000000000000000000000000000000000000000000000013FE5000000000000000000000000C663E7E7```
Yours will probably look very similar, only the bytes in the middle should change
Looking at the script you posted above, everything should work considering you have the correct port and baud rate (if you are on mac/linux, windows ports look different), the only necessary changes would be these lines I think
```
ecoded_data = data.decode('utf-8')  # Decode bytes to a string
print(decoded_data, end='')  # Print the data without a newline
```
This StackOverflow thread might help you a lot (bonus points if you pull that off without using a library, as that makes the integration lighter, and that's always good)
https://stackoverflow.com/questions/6624453/whats-the-correct-way-to-convert-bytes-to-a-hex-string-in-python-3

Once you are able to read the inverter and print that giant string of bytes to the terminal, please come back to me and I'll make sure to point you in the right direction to proceed in adding that code to the whole integration

Also, if you are interested, you can open an issue in the integration repo (https://github.com/LucasTor/FoxESS-T-series/issues), as that will make our communication easier!

**Reply by LucasTor** [reply]
@mattwilson1066 is your adapter this one? I will buy one so I can make some tests as well, so we can make sure it works the best way possible (bad thing it's gonna take a while to arrive, as I usually buy from China...)
If not, can you point me to the right one?
Thx!!
<img width="413" alt="Screenshot 2023-12-15 at 19 29 05" src="https://github.com/nathanmarlor/foxess_modbus/assets/25714552/2806f108-10f9-427f-b543-79bdf4f99829">

**Reply by mattwilson1066** [reply]
@LucasTor I actually got this one which I got from Amazon of Pii store.
https://www.waveshare.com/usb-to-rs485-b.htm

I'm hoping to spend an evening next week having a look at what you sent in your last message and then hopefully spend some time between Christmas and New Year trying to get it working. The problem with trying to do things in the evening after work is the Inverter goes into such a deep standby mode you'll get no Comms out of it until the sun comes out and it wakes up which makes it a little more difficult.
